mesh/exodus_writer.cpp

```diff
@@ -0,0 +1,1018 @@
+}
+
+void ExodusIIWriter::WriteBlockIDs()
+{
+   int block_dim;
+   DefineDimension(ExodusIILabels::EXODUS_NUM_BLOCKS_LABEL, _block_ids.size(),
+                   &block_dim);
+
+   DefineAndPutVar(ExodusIILabels::EXODUS_ELEMENT_BLOCK_IDS_LABEL, NC_INT, 1,
+                   &block_dim,
+                   _block_ids.data());
+}
+
+void ExodusIIWriter::WriteElementBlocks()
+{
+   GenerateExodusIIElementBlocks();
+
+   WriteNumElementBlocks();
+   WriteBlockIDs();
+
+   for (int block_id : _block_ids)
+   {
+      WriteElementBlockParameters(block_id);
+   }
+}
+
+char * ExodusIIWriter::GenerateLabel(const char * format, ...)
+{
+   va_list arglist;
+   va_start(arglist, format);
+
+   const int buffer_size = 100;
+
+   static char buffer[buffer_size];
+   int nwritten = vsnprintf(buffer, buffer_size, format, arglist);
+
+   bool ok = (nwritten > 0 && nwritten < buffer_size);
+   if (!ok)
+   {
+      MFEM_ABORT("Unable to write characters to buffer.");
+   }
+
+   va_end(arglist);
+   return buffer;
+}
+
+void ExodusIIWriter::WriteElementBlockParameters(int block_id)
+{
+   char * label{nullptr};
+
+   const std::vector<int> & block_element_ids = _element_ids_for_block_id.at(
+                                                   block_id);
+   const Element * front_element = _mesh.GetElement(block_element_ids.front());
+
+   // 1. Define number of elements in the block.
+   label = GenerateLabel("num_el_in_blk%d", block_id);
+
+   int num_el_in_blk_id;
+   DefineDimension(label, block_element_ids.size(),
+                   &num_el_in_blk_id);
+
+   // 2. Define number of nodes per element.
+   label = GenerateLabel("num_nod_per_el%d", block_id);
+
+   int num_node_per_el_id;
+   if (_mesh.GetNodes())
+   {
+      // Higher order. Get the first element from the block.
+      const FiniteElementSpace * fespace = _mesh.GetNodalFESpace();
+
+      auto & block_elements = _element_ids_for_block_id.at(block_id);
+
+      int first_element_id = block_elements.front();
+
+      Array<int> dofs;
+      fespace->GetElementDofs(first_element_id, dofs);
+
+      DefineDimension(label, dofs.Size(),
+                      &num_node_per_el_id);
+   }
+   else
+   {
+      DefineDimension(label, front_element->GetNVertices(),
+                      &num_node_per_el_id);
+   }
+
+   // 3. Define number of edges per element:
+   label = GenerateLabel("num_edg_per_el%d", block_id);
+
+   int num_edg_per_el_id;
+   DefineDimension(label, front_element->GetNEdges(),
+                   &num_edg_per_el_id);
+
+   // 4. Define number of faces per element.
+   label = GenerateLabel("num_fac_per_el%d", block_id);
+
+   int num_fac_per_el_id;
+   DefineDimension(label, front_element->GetNFaces(),
+                   &num_fac_per_el_id);
+
+   // 5. Define element node connectivity for block.
+   WriteNodeConnectivityForBlock(block_id);
+
+   // 6. Define the element type.
+   std::string element_type;
+
+   bool higher_order = (_mesh.GetNodes() != nullptr);
```

Comment:
We need to be a little more discerning here. At the very least we should confirm that the `Nodes` defined in the mesh do indeed correspond to a 2nd order H1 space (because that would seem to be the assumption made later on). We can then exit gracefully with a warning message if printing this particular type of mesh is not supported.

In the future we can perhaps address issues such as `Nodes` based on L2 spaces. We could even project higher order `Nodes` to 2nd order nodes as long as we print a warning message. These improvements do not need to be handled at this time but we should be clear about the limitations of the current implementation.

Reply:
Agreed. I've added a method to check the Nodal FESpace. It calls an abort if the space is not of the correct type. A warning would definitely be better although it will require more major changes to the writer class to do this elegantly.